src/components/ThumbnailComponent.js

```diff
@@ -0,0 +1,247 @@
+import {
+  EventHandlerService,
+  PDFLEvents,
+} from "../services/EventHandlerService";
+
+/**
+ * Component responsible for creating and displaying PDF thumbnail.
+ * The thumbnail contains all PDF pages which can be scrolled through and
+ * clicked.
+ * 
+ * @property {Object} components object that holds all DOM elements within this component
+ * @property {HTMLElement} components.toolbar element that represents this component
+ * @property {Object} pdfjsLib PDF.js library
+ * @property {boolean} isOpened state of component which can be either closed or opened
+ */
+class ThumbnailComponent {
+  components = {
+    thumbnail: document.querySelector("#thumbnail"),
+  };
+
+  /**
+   * Creates and initializes new thumbnail component. Sets component state to closed.
+   */
+  constructor() {
+    this.pdfjsLib = require("pdfjs-dist");
+    this.isOpened = false;
+
+    this.#registerEvents();
+  }
+
+  /**
+   * Registers event listeners to this component.
+   * @private
+   */
+  #registerEvents = () => {
+    EventHandlerService.subscribe(PDFLEvents.onReadNewFile, (pdfDoc) =>
+      this.#createNewThumbnail(pdfDoc)
+    );
+
+    EventHandlerService.subscribe(PDFLEvents.onToggleThumbnail, () =>
+      this.#toggle()
+    );
+
+    EventHandlerService.subscribe(PDFLEvents.onPageChanged, (pageNumber) =>
+      this.#setSelectedPage(document.querySelectorAll(".thumbnail-page-canvas")[pageNumber - 1])
+    );
+  };
+
+  /**
+   * Closes the thumbnail if it was opened and vice versa. Changes
+   * state accordingly.
+   * @private
+   */
+  #toggle = () => {
+    this.isOpened ? this.#close() : this.#open();
+
+    this.isOpened = !this.isOpened;
+  };
+
+  /**
+   * Displays thumbnail.
+   */
+  #open = () => {
+    this.components.thumbnail.classList.remove("hidden");
+  };
+
+  /**
+   * Hides thumbnail.
+   */
+  #close = () => {
+    this.components.thumbnail.classList.add("hidden");
+  };
+
+  /**
+   * Removes old and creates new thumbnail from PDF document data.
+   * @param {Uint8Array} pdfDoc PDF document data
+   */
+  #createNewThumbnail = async (pdfDoc) => {
+    this.#clear();
+
+    let document = await this.pdfjsLib.getDocument(pdfDoc).promise;
+    for (let pageNumber = 1; pageNumber <= document.numPages; pageNumber++)
+      await this.#createPage(await document.getPage(pageNumber), pageNumber);
+  };
+
+  /**
+   * Removes all page elements from thumbnail element.
+   */
+  #clear = () => {
+    this.components.thumbnail.innerHTML = "";
+  };
+
+  /**
+   * Creates new small PDF page in thumbnail. 
+   * @param {PDFPageProxy} page PDF page
+   * @param {int} num page number
+   */
+  #createPage = async (page, num) => {
+    let viewport = this.#createPageViewport(page);
+    let canvas = this.#createPageCanvas(viewport, num);
+
+    await page
```

Comment:
Why both await and promise?

Reply:
Try removing one or another and none of this will work, because this is not a standard js promise but a part of PDF.js library. getDocument function of PDF reader returns PDFDocumentLoadingTask which then has following "promise":
![image](https://user-images.githubusercontent.com/79309587/209599607-1f39d772-df3d-4816-a1a2-269fdc09c83f.png)
And then await is there because that promise has to be awaited, otherwise document is undefined.